vitessce/data_utils/anndata.py

```diff
@@ -0,0 +1,163 @@
+    # Only keep the subset of obsm and varm items required.
+    if obsm_keys is None:
+        obsm_keys = adata.obsm.keys()
+    new_obsm = {
+        obsm_key: optimize_arr(adata.obsm[obsm_key])
+        for obsm_key
+        in obsm_keys
+    }
+    if varm_keys is None:
+        varm_keys = adata.varm.keys()
+    new_varm = {
+        varm_key: optimize_arr(adata.varm[varm_key])
+        for varm_key
+        in varm_keys
+    }
+    if layer_keys is None:
+        layer_keys = adata.layers.keys()
+    new_layers = {
+        layer_key: optimize_arr(adata.layers[layer_key])
+        for layer_key
+        in layer_keys
+    }
+    adata = AnnData(X=new_X, obs=new_obs, var=new_var, obsm=new_obsm, varm=new_varm, layers=new_layers)
+    return adata
+
+
+def to_memory(arr):
+    if type(arr).__name__ == "SparseDataset":
+        return arr.to_memory()
+    return arr
+
+
+# Convert a sparse array to dense.
+def to_dense(arr):
+    if issparse(arr):
+        return arr.todense()
+    return arr
+
+
+# Convert an array to uint8 dtype.
+def to_uint8(arr, norm_along=None):
+    # Re-scale the gene expression values between 0 and 255 (one byte ints).
+    if norm_along is None:
+        norm_arr = arr
+    elif norm_along == "global":
+        arr *= 255.0 / arr.max()
+        norm_arr = arr
+    elif norm_along == "var":
+        # Normalize along gene axis
+        arr = to_dense(arr)
+        num_cells = arr.shape[0]
+        min_along_genes = arr.min(axis=0)
+        max_along_genes = arr.max(axis=0)
+        range_per_gene = max_along_genes - min_along_genes
+        ratio_per_gene = 255.0 / range_per_gene
+
+        norm_arr = np.multiply(
+            (arr - np.tile(min_along_genes, (num_cells, 1))),
+            np.tile(ratio_per_gene, (num_cells, 1))
+        )
+    elif norm_along == "obs":
+        # Normalize along cell axis
+        arr = to_dense(arr)
+        num_genes = arr.shape[1]
+        min_along_cells = arr.min(axis=1)
+        max_along_cells = arr.max(axis=1)
+        range_per_cell = max_along_cells - min_along_cells
+        ratio_per_cell = 255.0 / range_per_cell
+
+        norm_arr = np.multiply(
+            (arr.T - np.tile(min_along_cells, (num_genes, 1))),
+            np.tile(ratio_per_cell, (num_genes, 1))
+        ).T
+    else:
+        raise ValueError("to_uint8 received unknown norm_along value")
+    return norm_arr.astype('u1')
+
+
+# Sort the var index after hierarchical clustering.
+def sort_var_axis(adata_X, orig_var_index):
+    gexp_arr = to_dense(adata_X)
+
+    # Perform hierarchical clustering along the genes axis.
+    Z = scipy.cluster.hierarchy.linkage(gexp_arr.T, method="ward")
+
+    # Get the hierarchy-based ordering of genes.
+    leaf_index_list = scipy.cluster.hierarchy.leaves_list(Z)
+    leaf_list = orig_var_index[leaf_index_list].tolist()
+
+    return leaf_list
+
+
+# Convert an (x, y) coordinate to a polygon (diamond) with a given radius.
+def to_diamond(x, y, r):
```

Comment:
I thought that our cells API makes diamonds by default when no polygon is present?  Could have sworn vitessce did this....

Reply:
Yes it does but I would prefer to force people to explicitly make the diamonds at data processing time if they want to fake the cell segmentation process. Ideally a real cell segmentation algorithm should be used. It is removed from the client side in https://github.com/vitessce/vitessce/pull/1261